## Wstępna analiza

Wczytujemy plik csv używając biblioteki pandas jako DataFrame, upewniamy się że widoczne będą wszystkie kolumny z danymi, analizujemy wielkość i naturę zbioru danych.

In [2]:
import pandas as pd
df = pd.read_csv('data/final_train.csv')
pd.set_option('display.max_columns', None)

In [3]:
df.shape

(11086, 573)

In [4]:
df.sample(6)

Unnamed: 0            Activity  angle(X,gravityMean)  \
6364         6364             SITTING             -1.568834   
324           324              LAYING             -1.548333   
5647         5647    WALKING_UPSTAIRS             -1.568669   
7163         7163             WALKING             -1.570576   
9520         9520  WALKING_DOWNSTAIRS             -1.573969   
10119       10119             WALKING             -1.572876   

       angle(Y,gravityMean)  angle(Z,gravityMean)  \
6364              -0.566471             -2.425402   
324               -0.139628             -2.380889   
5647              -0.612620             -2.488338   
7163              -0.951505             -2.463952   
9520              -0.712497             -2.444000   
10119             -0.787974             -2.469375   

       angle(tBodyAccJerkMean),gravityMean)  angle(tBodyAccMean,gravity)  \
6364                              -1.889798                     0.081603   
324                               -1.915239                     0.067613   
5647                              -1.751117                     0.218134   
7163                              -1.774724                    -0.413562   
9520                              -1.913643                     0.317868   
10119                             -1.983042                     0.230050   

       angle(tBodyGyroJerkMean,gravityMean)  angle(tBodyGyroMean,gravityMean)  \
6364                              -0.655480                          4.487523   
324                               -0.887342                          6.007361   
5647                               0.280294                          7.953403   
7163                              -1.659248                          4.285869   
9520                              -1.429880                          1.903187   
10119                              0.160301                          1.867394   

       energy-mean()  fBodyAcc-bandsEnergy()-1,16  \
6364       -8.831272                     3.623740   
324        -5.793913                     3.621984   
5647       -0.679894                     2.512880   
7163      -11.962755                     2.729030   
9520       -4.042786                     1.254758   
10119      -8.712189                     2.438280   

       fBodyAcc-bandsEnergy()-1,16.1  fBodyAcc-bandsEnergy()-1,16.2  \
6364                        6.156064                       2.496019   
324                         6.158630                       2.495276   
5647                        5.740178                       2.803435   
7163                        5.066370                       2.721687   
9520                        5.465639                       2.598826   
10119                       5.714800                       2.857791   

       fBodyAcc-bandsEnergy()-1,24  fBodyAcc-bandsEnergy()-1,24.1  \
6364                      5.777292                       3.306882   
324                       5.776843                       3.314383   
5647                      5.484726                       2.102368   
7163                      5.523828                      -0.039186   
9520                      5.190757                       1.212844   
10119                     5.438408                       1.789598   

       fBodyAcc-bandsEnergy()-1,24.2  fBodyAcc-bandsEnergy()-1,8  \
6364                       -1.352867                   -1.537868   
324                        -1.352860                   -1.537483   
5647                       -1.355854                   -1.416568   
7163                       -1.355174                   -1.379305   
9520                       -1.354082                   -1.050899   
10119                      -1.357317                   -1.349333   

       fBodyAcc-bandsEnergy()-1,8.1  fBodyAcc-bandsEnergy()-1,8.2  \
6364                       2.844830                      8.617460   
324                        2.845718                      8.618923   
5647                       2.527046                   

W trakcie analizy dochodzimy do kilku wniosków, mianowicie:

In [5]:
pd.unique(df['Activity'])

array(['STANDING', 'WALKING_UPSTAIRS', 'LAYING', 'WALKING', 'SITTING',
       'WALKING_DOWNSTAIRS'], dtype=object)

1. Mamy 6 aktywności, które nasz model ma rozpoznawać,

In [6]:
df.subject.nunique()

30

2. W badaniu brało udział 30 uczestników,

3. Do trenowania modelu klasyfikacji definitywnie nie potrzebujemy wszystkich cech. Skupiamy się na tych, które zawierają średnią i odchylenie standardowe.

## Oczyszczanie danych

Zanim zaczniemy czyścić zbiór, wyciągamy z niego dane o uczestnikach i aktywnościach, to przyda nam się potem.

In [7]:
subject = df['subject']
activity = df['Activity']

Na początku wybieramy wszystkie kolumny zawierające średnią lub odchylenie standardowe wszystkich z cech.

In [8]:
mean_cols = [col for col in df.columns if 'mean' in col or 'Mean' in col]
std_cols = [col for col in df.columns if 'std' in col or 'Std' in col]

df = pd.concat([df[std_cols], df[mean_cols]], axis = 1)
df.shape

(11086, 87)

In [9]:
573-87, (573-87)/573

(486, 0.8481675392670157)

W ten sposób udało nam się pozbyć 486 cech, co stanowi prawie 85% całego zbioru.

Aktualnie nasz zbiór wygląda tak:

In [10]:
df.sample(4)

fBodyAcc-std()-X  fBodyAcc-std()-Y  fBodyAcc-std()-Z  \
10496          0.016032         -2.836052          3.993116   
5901           0.782542         -1.315487          4.206194   
6988          -1.283255         -3.161357          3.870946   
502           -0.224699         -3.077772          3.630931   

       fBodyAccJerk-std()-X  fBodyAccJerk-std()-Y  fBodyAccJerk-std()-Z  \
10496             -5.066653              1.933033              2.093067   
5901              -4.035635              0.271185              2.355980   
6988              -6.250629              2.219342              1.475529   
502               -5.089513              1.843529              1.611339   

       fBodyAccMag-std()  fBodyBodyAccJerkMag-std()  \
10496          -1.513092                  -2.890308   
5901           -0.838646                  -3.015638   
6988           -2.804752                  -2.824515   
502            -1.900339                  -2.911878   

       fBodyBodyGyroJerkMag-std()  fBodyBodyGyroMag-std()  fBodyGyro-std()-X  \
10496                   -6.050516                4.246825          -0.914943   
5901                    -6.016085                4.841528           3.323182   
6988                    -6.081195                4.650028           1.565835   
502                     -6.058978                4.639054           2.182885   

       fBodyGyro-std()-Y  fBodyGyro-std()-Z  tBodyAcc-std()-X  \
10496           0.755810           4.426131          6.058817   
5901            0.806858           4.685420          6.341033   
6988            0.710725           4.134100          5.634211   
502             0.649238           3.913713          5.993834   

       tBodyAcc-std()-Y  tBodyAcc-std()-Z  tBodyAccJerk-std()-X  \
10496          3.764960          2.950739              5.401974   
5901           3.881736          3.440171              6.304277   
6988           3.733998          2.540716              4.571304   
502            3.750150          2.136298              5.533845   

       tBodyAccJerk-std()-Y  tBodyAccJerk-std()-Z  tBodyAccJerkMag-std()  \
10496             -2.056362              3.721760               2.917945   
5901              -0.530593              3.919964               2.316887   
6988              -2.307205              3.291088               3.347972   
502               -2.005583              3.383666               2.814718   

       tBodyAccMag-std()  tBodyGyro-std()-X  tBodyGyro-std()-Y  \
10496           0.359691           1.672970           4.336518   
5901            1.598607           5.369631           5.607410   
6988           -1.509697           3.624113           3.071137   
502            -0.035756           4.241027           1.836998   

       tBodyGyro-std()-Z  tBodyGyroJerk-std()-X  tBodyGyroJerk-std()-Y  \
10496           1.470095              -3.506829              -0.551104   
5901            1.774585              -4.165298              -0.590117   
6988            1.183452              -3.191312              -0.471662   
502             1.026470              -3.557145              -0.481587   

       tBodyGyroJerk-std()-Z  tBodyGyroJerkMag-std()  tBodyGyroMag-std()  \
10496               0.853528               -2.485222            1.621000   
5901                0.852435               -2.666552            5.772904   
6988                0.854424               -2.302259            4.127234   
502                 0.854039               -2.402722            4.109122   

       tGravityAcc-std()-X  tGravityAcc-std()-Y  tGravityAcc-std()-Z  \
10496             3.691703             1.262103             4.041029   
5901              3.714004             1.424940             4.095446   
6988              3.679519             1.291738             4.036252   
502               3.701963             1.320733             4.027137   

       tGravityAccMag-std()  angle(X,gravityMean)  angle(Y,gravityMean)  \
10496              0.517117             -1.575067             -0.713835   
5

Dodatkowo zdecydowaliśmy się pozbyć kolumn, które zawierają słowa kluczowe: 'angle', 'energy' i 'meanFreq'

In [11]:
angle_cols = [col for col in df.columns if 'angle' in col or 'Angle' in col]
energy_cols = [col for col in df.columns if 'energy' in col or 'Energy' in col]
meanFreq_cols = [col for col in df.columns if 'meanFreq' in col]

drop_cols = angle_cols + energy_cols + meanFreq_cols
df = df.drop(drop_cols, axis=1)

In [12]:
df.shape

(11086, 66)

Ostatecznie nasz zbiór ma 66 cech.

Dodajemy kolumny uczestnika i aktywności.

In [13]:
df.insert(0, 'Activity', activity)
df.insert(0, 'Subject', subject)

Pozbywamy się niekompletnych wierszy.

In [14]:
df = df.dropna()

Dane, które użyliśmy do przetestowania modelu

In [15]:
df.head(10)

Subject            Activity  fBodyAcc-std()-X  fBodyAcc-std()-Y  \
0        30            STANDING          0.756296         -1.586674   
1         9    WALKING_UPSTAIRS         -0.537596         -2.784092   
2        28              LAYING          0.755296         -1.283612   
3        25             WALKING          0.149588         -2.796041   
4        28              LAYING          0.755296         -1.283612   
5        23             SITTING          0.664783         -2.040734   
6        27              LAYING          0.802888         -1.258179   
7        14            STANDING          0.794411         -1.344356   
9         7  WALKING_DOWNSTAIRS         -1.309521         -2.932957   
10       17             WALKING         -0.431861         -2.956632   

    fBodyAcc-std()-Z  fBodyAccJerk-std()-X  fBodyAccJerk-std()-Y  \
0           4.141758             -4.176057              0.585073   
1           3.677794             -5.817531              1.293797   
2           4.217478             -4.063730              0.303414   
3           3.922191             -4.776575              1.763192   
4           4.217478             -4.063730              0.303414   
5           4.042778             -4.141761              0.414488   
6           4.220286             -4.023362              0.263590   
7           4.215034             -4.022720              0.291862   
9           3.863861             -6.394034              1.706218   
10          3.794916             -5.436230              2.108401   

    fBodyAccJerk-std()-Z  fBodyAccMag-std()  fBodyBodyAccJerkMag-std()  \
0               2.245282          -1.000675                  -2.979941   
1               1.616815          -2.166122                  -2.906000   
2               2.333802          -0.806943                  -3.011841   
3               1.876670          -1.633627                  -2.920310   
4               2.333802          -0.806943                  -3.011841   
5               2.316583          -1.467383                  -2.994722   
6               2.358539          -0.768377                  -3.016589   
7               2.360033          -0.830130                  -3.015687   
9               1.549443          -3.221167                  -2.809464   
10              1.639118          -1.476475                  -2.884412   

    fBodyBodyGyroJerkMag-std()  fBodyBodyGyroMag-std()  fBodyGyro-std()-X  \
0                    -6.033122                4.781842           2.565681   
1                    -6.062388                4.653310           1.148811   
2                    -6.017739                4.845153           3.302522   
3                    -6.048324                4.737144           1.710101   
4                    -6.017739                4.845153           3.302522   
5                    -6.027634                4.785031           3.007326   
6                    -6.015245                4.852861           3.342725   
7                    -6.016443                4.843876           3.254727   
9                    -6.100774                4.680560           1.896524   
10                   -6.080514                4.762033           1.975335   

    fBodyGyro-std()-Y  fBodyGyro-std()-Z  tBodyAcc-std()-X  tBodyAcc-std()-Y  \
0            0.795967           4.632040          6.325173          3.860434   
1            0.660377           3.794268          5.880477          3.775318   
2            0.807615           4.705391          6.331799          3.883619   
3            0.746790           4.400935          6.119588          3.769096   
4            0.807615           4.705391          6.331799          3.883619   
5            0.796710           4.558747          6.302001          3.832121   
6            0.813807           4.718742          6.347577          3.885898   
7            0.812325           4.695774          6.345157          3.879934   
9            0.726239           4.169236          5.663822          3.759751   
10        

In [16]:
df.shape

(9826, 68)

Ostatecznie zdecydowaliśmy się zgrupować wszystkich uczestników badania razem z aktywnościami, w których brali udział

In [17]:
grouped = df.groupby(['Subject', 'Activity'])

Wyciągamy średnią z wszystkich cech, dla każdego uczestnika względem każdej z badanych aktywności.

In [18]:
df = grouped.mean().reset_index()

I tak prezentują się nasze zgrupowane oczyszczone dane, które użyliśmy do trenowania modelu:

In [19]:
df.head(10)

Subject            Activity  fBodyAcc-std()-X  fBodyAcc-std()-Y  \
0        1              LAYING          0.772614         -1.312344   
1        1             SITTING          0.798087         -1.307646   
2        1            STANDING          0.802282         -1.294112   
3        1             WALKING         -0.424408         -3.359302   
4        1  WALKING_DOWNSTAIRS         -0.938943         -3.193383   
5        1    WALKING_UPSTAIRS         -0.358470         -3.207309   
6        2              LAYING          0.775048         -1.254903   
7        2             SITTING          0.784312         -1.341305   
8        2            STANDING          0.789539         -1.324953   
9        2             WALKING         -0.151261         -2.896934   

   fBodyAcc-std()-Z  fBodyAccJerk-std()-X  fBodyAccJerk-std()-Y  \
0          4.192530             -4.032061              0.269285   
1          4.214334             -4.022398              0.260490   
2          4.213188             -4.021302              0.268362   
3          3.828758             -5.753965              2.741275   
4          3.840144             -5.907574              2.474086   
5          3.625047             -5.119551              1.600063   
6          4.219320             -4.029690              0.263934   
7          4.200123             -4.032743              0.281445   
8          4.197566             -4.049089              0.304747   
9          3.921515             -5.403117              2.481886   

   fBodyAccJerk-std()-Z  fBodyAccMag-std()  fBodyBodyAccJerkMag-std()  \
0              2.358421          -0.846795                  -3.015795   
1              2.365289          -0.804229                  -3.016613   
2              2.364522          -0.805330                  -3.016604   
3              1.604912          -2.147099                  -2.845857   
4              1.407951          -2.649033                  -2.837071   
5              1.937368          -2.063600                  -2.928284   
6              2.361821          -0.785951                  -3.016474   
7              2.358394          -0.868957                  -3.015060   
8              2.350316          -0.848490                  -3.012764   
9              1.742390          -1.741868                  -2.858549   

   fBodyBodyGyroJerkMag-std()  fBodyBodyGyroMag-std()  fBodyGyro-std()-X  \
0                   -6.016373                4.840281           3.258072   
1                   -6.015543                4.839362           3.315200   
2                   -6.015761                4.848332           3.322622   
3                   -6.102188                4.649459           1.876334   
4                   -6.111458                4.659888           1.507965   
5                   -6.056443                4.606615           2.011382   
6                   -6.015844                4.843700           3.327517   
7                   -6.016521                4.841547           3.314211   
8                   -6.018500                4.840633           3.278964   
9                   -6.078999                4.746789           2.091507   

   fBodyGyro-std()-Y  fBodyGyro-std()-Z  tBodyAcc-std()-X  tBodyAcc-std()-Y  \
0           0.812411           4.702018          6.337974          3.882237   
1           0.812666           4.674744          6.346444          3.882630   
2           0.812209           4.704193          6.347743          3.883326   
3           0.654530           4.139078          5.906645          3.724877   
4           0.678322           3.915508          5.741504          3.731460   
5           0.628777           4.280835          5.962865          3.746696   
6           0.811658           4.683152          6.339253          3.885898   
7           0.810556           4.680929          6.341413          3.879928   
8           0.809204           4.688025          6.342291          3.880793   
9           0.737212           4.254019          5.998372      

In [20]:
df.shape

(180, 68)

## Model

Do implementacji modelu użyliśmy biblioteki TensorFlow 

### Definicja modelu

In [ ]:
# build model: 66 -> 150 -> 80 -> 6
model = tf.kras.models.Sequential([
        keras.layers.Dense(150, input_dim=66, activation="relu"),
        keras.layers.Dense(80, activation="relu"),
        keras.layers.Dense(6, activation="softmax")
])

Jest to prosty model. Używa dwóch warstw ukrytych z funkcjami aktywacji ReLU. 

Na ostatniej warstwie jest użyty softmax. Dzięki niej, wartości otrzymane są jak najbardziej czytelne. Sumują się one w przybliżeniu do 1 oraz wartość największa mówi nam, która z czynności jest prawdopodobnie wykonywana.

Dzięki temu możemy w bardziej skomplikowanych implementacjach użyć na przykład drugiej największej wartości.

### Kompilacja modelu

In [ ]:
model.compile(optimizer="adam", loss="MSE", metrics=['accuracy'])

### Trenowanie

Następnym krokiem było trenowanie. Użyliśmy do tego zgrupowanych oczyszczonych danych (6 aktywności na 30 osób - 180 pomiarów)

In [66]:
model.fit(X, Y, epochs=200)

Epoch 1/200
6/6 [==============================] - 0s 609us/step - loss: 0.1620 - accuracy: 0.2444
Epoch 2/200
6/6 [==============================] - 0s 834us/step - loss: 0.1320 - accuracy: 0.3889
Epoch 3/200
6/6 [==============================] - 0s 500us/step - loss: 0.1183 - accuracy: 0.4167
Epoch 4/200
6/6 [==============================] - 0s 834us/step - loss: 0.1049 - accuracy: 0.6444
Epoch 5/200
6/6 [==============================] - 0s 668us/step - loss: 0.0992 - accuracy: 0.6389
Epoch 6/200
6/6 [==============================] - 0s 832us/step - loss: 0.0966 - accuracy: 0.5833
Epoch 7/200
6/6 [==============================] - 0s 499us/step - loss: 0.0904 - accuracy: 0.6667
Epoch 8/200
6/6 [==============================] - 0s 668us/step - loss: 0.0896 - accuracy: 0.6056
Epoch 9/200
6/6 [==============================] - 0s 667us/step - loss: 0.0990 - accuracy: 0.5944
Epoch 10/200
6/6 [==============================] - 0s 666us/step - loss: 0.0926 - accuracy: 0.6167
Epoch 11/

6/6 [==============================] - 0s 835us/step - loss: 0.0147 - accuracy: 0.9333
Epoch 165/200
6/6 [==============================] - 0s 501us/step - loss: 0.0179 - accuracy: 0.9111
Epoch 166/200
6/6 [==============================] - 0s 832us/step - loss: 0.0155 - accuracy: 0.9333
Epoch 167/200
6/6 [==============================] - 0s 833us/step - loss: 0.0160 - accuracy: 0.9389
Epoch 168/200
6/6 [==============================] - 0s 499us/step - loss: 0.0236 - accuracy: 0.8889
Epoch 169/200
6/6 [==============================] - 0s 667us/step - loss: 0.0193 - accuracy: 0.9167
Epoch 170/200
6/6 [==============================] - 0s 1ms/step - loss: 0.0168 - accuracy: 0.9389
Epoch 171/200
6/6 [==============================] - 0s 1ms/step - loss: 0.0149 - accuracy: 0.9500
Epoch 172/200
6/6 [==============================] - 0s 834us/step - loss: 0.0149 - accuracy: 0.9389
Epoch 173/200
6/6 [==============================] - 0s 501us/step - loss: 0.0131 - accuracy: 0.9500
Epoch 17

### Dokładność

Po przetrenowaniu modelu na wcześniejszych danych, sprawdziliśmy dokładność modelu

In [25]:
model.evaluate(X_2, Y_2, verbose=1)


 model evaluation:
308/308 [==============================] - 0s 404us/step - loss: 0.0198 - accuracy: 0.9232


[0.019759194925427437, 0.9231630563735962]

Osiągnęliśmy dokładność rzędu 92% na oczyszczonych danych (wyeliminowanie wierszy z NaNami (9826 pomiarów))

Gotowy model zapisaliśmy do cp.ckpt